Tour Kit: Allow CSV and array for custom classes

arcangelini opened the pull request:
## Changes proposed in this Pull Request
As suggested in the GH issue #59710 this allows both the frame and step elements to accept CSV and arrays for the `classname` option.

To simplify this I brought the logic for both elements into the frame and passed the step classes down to the step. This allowed me to create a small function to filter and split any CSV without having to duplicate the same code in the frame and step elements.

*I am not 100% sure I updated the config* properly so please be sure to double-check that : ) I am not seeing any TS errors in my testing.

## Screenshots
![Screen Capture on 2022-01-19 at 13-29-47](https://user-images.githubusercontent.com/33258733/150132546-81e68353-30cd-48f1-91ec-a7e864037803.gif)

## Testing instructions

1. Checkout branch
2. Go to `apps/editing-toolkit/` and run `yarn dev --sync` to push changes to your sandbox
3. Edit `apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-steps.js` and set `className` option to be an array or CSV and save.
4. Do the same to `apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js`
5. Open the Welcome Guide in your sandbox site and check to make sure the classes appear properly

Related to #59710

chriskmnds commented on packages/tour-kit/src/components/tour-kit-frame.tsx:
```diff
@@ -183,13 +183,29 @@ const TourKitFrame: React.FunctionComponent< Props > = ( { config } ) => {
+		'tour-kit-step',
+		`is-step-${ currentStepIndex }`,
+		config.options?.className
+			? classSplit( config.options?.className )?.map( ( value ) => value + '_step' )
```
Should be `__step`.

However, I'm having second thoughts about the inclusion of these "magic" classes. They sort of lost their appeal to me, especially now that we handle multiple classes from the root component. Imagine a scenario where we pass `foo-tour` and `is-foo-tour-modifier` to the parent, we end up with `foo-tour__step` and `is-foo-tour-modifier__step` in the step. Feels complex, hard to communicate, and hard to be trusted by the consumer. I would vote to remove this part altogether.

What do you think? If removing, we can just send [wpcom-editor-welcome-tour__step](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/style-tour.scss#L220) along with `is-with-extra-padding` now that we handle multiple classes for steps.

arcangelini replied:
Cool, I probably should have raised this concern when I put this together. I agree the classes get a bit excessive when we add `__step` to all the parent classes. I will clean this up 🕺

chriskmnds replied:
Cool! Thanks, and please feel free to raise concerns at any point. 😃

chriskmnds reviewed (commented):
Amazing! Thanks for the great work @arcangelini. Just a comment below, but otherwise looks great :-)

escapemanuele reviewed (changes requested):
Thank you @arcangelini, great job here! 

There is one more thing to consider though, after cleaning up the step CSS classes (no more bloating with a lot of ...__step classes, well done), we should also change the [tour-steps](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-steps.js#L88-L90), reflecting the changes you made.
In this case, we'd need to add the two CSS classes used in [style-tour](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/style-tour.scss#L220-L221) to add the extra padding 😄

arcangelini commented:
Thanks @escapemanuele I completely missed that part of the comments. I made a mess of the branch and had to pull some fancy maneuvers to get it cleaned up. In the process, I lost your suggested changes and can't resolve them. Anyways, the changes should be committed and good to go.

chriskmnds commented on packages/tour-kit/src/components/tour-kit-step.tsx:
```diff
@@ -15,20 +11,14 @@ const TourKitStep: React.FunctionComponent< Props > = ( {
 	setInitialFocusedElement,
 	onGoToStep,
 } ) => {
-	const classNames = classnames(
```
Let's maybe instead move `classSplit` above into a utils folder and reuse here to keep concerns isolated (so `step` owns its class definitions). Frame is owning a little too much as is already.

arcangelini replied:
Should I make a utils directory or just a utils file? Should the dir/file go in the components directory or should it go directly into src?

Let me know if we have some more stuff that you think can be pulled out I can add them in there too.

chriskmnds replied:
Good question. Well for now, given the size of the package, we can keep in a single file, thinking under `src` (or a `src/utils/index.ts` one if you prefer). Maybe add a test to go along?

We can keep an eye out for reusable utilities and see if down the line we can split out.

arcangelini commented:
OK, scrapped the function altogether in favor of a simpler approach. The original issue the function was solving was handling `classes` being passed either as a string or array. By adding `toString` before the split it accounts for both cases. Tested again and all is working properly.

Removed the function and put the `classes` processing back into the respective elements. We shouldn't need the `utils` file anymore, but please correct me if I am wrong.

Thanks for y'alls patience with my learning process haha.

escapemanuele reviewed (commented):
Thanks for working on this @arcangelini, great job! 

I tested it and it works nicely!

What I'm thinking though is that it would still be nice to have `options?.className?.toString().split( ',' )` in one place (like in a utility file as we said), as it may be a little complicated to understand reading the code. Well, maybe not too complicated, but that would also give us the opportunity to add a little test to it! 😄 

What do you think?

arcangelini commented:
Thanks @escapemanuele I refactored this and updated the prop name from `className` to `classes` to make it easier to understand now that it is able to accept multiple values.

escapemanuele commented:
Thanks again, @arcangelini, good job extracting the utility function! Just a couple of considerations, after that I think we'll be ready to go (after a rebase 😄 ).

I think it'd be better to maintain `className,` to make what we're talking about more understandable since React uses `className` for that purpose. But you're right; we may want to reflect the new multiple values possibility! What about using `classNames`? 

It'd be nice to have a little test for this utility function, just to test edge cases! You can add a test file in `/src/test/utils.ts` I think 🤔  Feel free to ask for help!

arcangelini committed:
Squashed commit of the following:

commit 81814b1db0d70336e579ea8cb86562e786c5ad88
Author: Tony Arcangelini <tony@arcangelini.com>
Date:   Tue Jan 25 21:34:08 2022 +0100

    Refactor to add utils - update className to classes

commit e721142bd27defcdd084855917d5edf117d0b539
Author: Tony Arcangelini <tony@arcangelini.com>
Date:   Sat Jan 22 16:27:51 2022 +0100

    Remove classSpit function in favor of inline approach

commit c00c5e391fa4f3c2a8aec5d1a6406d2abfe15c31
Author: Tony Arcangelini <tony@arcangelini.com>
Date:   Fri Jan 21 12:19:04 2022 +0100

    Add step class to welcome-tour__step and remove __step class from child

commit d92ff20f9ef164d9cf9e7c964456a85f755396e8
Author: Tony Arcangelini <tony@arcangelini.com>
Date:   Wed Jan 19 13:31:36 2022 +0100

    Allow CSV and array for classNames

arcangelini committed:
Chnage classes back to className

arcangelini commented:
You are right, I think it would be best to keep with the React naming of this. I switched it back to `className` and updated the branch.

As for the testing, I am going to spend some time working on this to better understand how to do this. Would it be ok to add the test in a second PR? or would it be better to wait to push this with the included tests?

arcangelini committed:
Add testing and updated class prop to classNames

escapemanuele commented on packages/tour-kit/README.md:
```diff
@@ -14,7 +14,7 @@ A tour is made up of the following components:
   - a set of optional reference elements selectors for rendering a step near
   - a set of options:
-	- className: optional custom CSS class that will be applied to the step
+	- classNames: optional custom CSS class that will be applied to the step
```
```suggestion
	- classNames: optional custom CSS classes that will be applied to the step
```

escapemanuele commented on packages/tour-kit/src/utils/index.ts:
```diff
@@ -0,0 +1,13 @@
+ */
+export function classParser( classes?: string | string[] ): string[] | null {
+	if ( classes?.length ) {
+		return classes?.toString().split( ',' );
```
I think we can skip the optional chaining here, as we already verified that `classed` contains something 👍

arcangelini replied:
I added that to make sure the function returns `null` in the event that *classes* is either `null` or an empty array.

Without `?.length` the function returns `['']` if given an empty array
Without making it optional the function errors when given a `null` value

Thoughts?

escapemanuele replied:
Oh I was referring to `return classes?.toString().split( ',' );`, the optional chaining there! 
Inside the if it's perfectly ok

arcangelini replied:
Ahhhh sorry haha misread that

escapemanuele reviewed (changes requested):
Great job here @arcangelini! 

Only two minor changes and after that I think we'll be ready to deploy 😄

arcangelini committed:
Update packages/tour-kit/README.md

Co-authored-by: Emanuele Buccelli <escapemanuele@gmail.com>

arcangelini committed:
Remove optional chaining on return

escapemanuele reviewed (approved):
Ship it! 🚀 :shipit:
